plugins/outputs/cloudwatch/aggregator_test.go

```diff
@@ -21,118 +21,110 @@ var metricName = "metric1"
-	close(shutdownChan)
 }
 
 func TestAggregator_NotDurationType(t *testing.T) {
-	metricChan, shutdownChan, aggregator := testPreparation()
+	metricChan, shutdownChan, agg := testPreparation()
+	defer testCleanup(t, metricChan, shutdownChan)
 	//aggregation key found, but no time.Duration type for the value
 	tags := map[string]string{"d1key": "d1value", "d2key": "d2value", aggregationIntervalTagKey: "1"}
 	fields := map[string]interface{}{"value": 1}
 	timestamp := time.Now()
 	m, _ := metric.New(metricName, tags, fields, timestamp)
 
-	aggregator.AddMetric(m)
+	agg.AddMetric(m)
 	select {
 	case aggregatedMetric := <-metricChan:
 		assert.False(t, aggregatedMetric.HasTag(aggregationIntervalTagKey))
 		assert.Equal(t, m, aggregatedMetric)
 	default:
 		assert.Fail(t, "Got no metrics")
 	}
-
-	assertNoMetricsInChan(t, metricChan)
-	close(shutdownChan)
 }
 
 func TestAggregator_ProperAggregationKey(t *testing.T) {
-	metricChan, shutdownChan, aggregator := testPreparation()
+	metricChan, shutdownChan, agg := testPreparation()
+	defer testCleanup(t, metricChan, shutdownChan)
 	//normal proper aggregation key found
 	aggregationInterval := 1 * time.Second
 	tags := map[string]string{"d1key": "d1value", "d2key": "d2value", aggregationIntervalTagKey: aggregationInterval.String()}
 	fields := map[string]interface{}{"value": 1}
 	timestamp := time.Now()
 	m, _ := metric.New(metricName, tags, fields, timestamp)
 
-	aggregator.AddMetric(m)
+	agg.AddMetric(m)
 	assertNoMetricsInChan(t, metricChan)
 	assertMetricContent(t, metricChan, aggregationInterval * 2, m, expectedFieldContent{"value", 1, 1, 1, 1, "",
 		[]float64{1.0488088481701516}, []float64{1}})
-
-	assertNoMetricsInChan(t, metricChan)
-	close(shutdownChan)
 }
 
 func TestAggregator_MultipleAggregationPeriods(t *testing.T) {
-	metricChan, shutdownChan, aggregator := testPreparation()
+	metricChan, shutdownChan, agg := testPreparation()
+	defer testCleanup(t, metricChan, shutdownChan)
 	//multiple metrics with multiple aggregation period, some aggregation period are the same, some are not.
 	timestamp := time.Now()
 	aggregationInterval := 1 * time.Second
 
 	tags := map[string]string{"d1key": "d1value", "d2key": "d2value", aggregationIntervalTagKey: aggregationInterval.String()}
 	fields := map[string]interface{}{"value": 1}
 	m, _ := metric.New(metricName, tags, fields, timestamp)
-	aggregator.AddMetric(m)
+	agg.AddMetric(m)
 
 	fields = map[string]interface{}{"value": 2}
 	m, _ = metric.New(metricName, tags, fields, timestamp)
-	aggregator.AddMetric(m)
+	agg.AddMetric(m)
 
 	fields = map[string]interface{}{"value": 3}
 	m, _ = metric.New(metricName, tags, fields, timestamp)
-	aggregator.AddMetric(m)
+	agg.AddMetric(m)
 
-	tags = map[string]string{"d1key": "d1value", "d2key": "d2value", aggregationIntervalTagKey: (2 * aggregationInterval).String()}
+	tags = map[string]string{"d1key": "d1value", "d2key": "d2value", aggregationIntervalTagKey: (3 * aggregationInterval).String()}
```

Comment:
My theory is that the flakiness stems from the aggregation interval. Originally, this was set to 2x the aggregation interval, and the test would wait 2x aggregation interval for a metric. Because of the timing, this sometimes missed the mark. A PR later attempted to update the aggregation interval itself, which did not alleviate the flakiness. 

Then a later PR than that increased the wait time for the tests, which alleviated the flakiness. Based on this, I believe that so long as the wait time > aggregation interval, this will significantly increase the effectiveness of the unit test.

Reply:
https://github.com/aws/amazon-cloudwatch-agent/runs/5294867347?check_suite_focus=true disproves my theory. Will keep looking